driver/razermouse_driver.c

```diff
@@ -4666,6 +4686,139 @@ static void razer_mouse_init(struct razer_mouse_device *dev, struct usb_interfac
+    case USB_DEVICE_ID_RAZER_MAMBA_WIRELESS:
+    case USB_DEVICE_ID_RAZER_MAMBA_WIRED:
+    case USB_DEVICE_ID_RAZER_OUROBOROS:
+    case USB_DEVICE_ID_RAZER_NAGA_EPIC_CHROMA:
+    case USB_DEVICE_ID_RAZER_NAGA_EPIC_CHROMA_DOCK:
+    case USB_DEVICE_ID_RAZER_NAGA_PRO_WIRELESS:
+    case USB_DEVICE_ID_RAZER_NAGA_PRO_WIRED:
+    case USB_DEVICE_ID_RAZER_MAMBA_WIRELESS_RECEIVER:
+    case USB_DEVICE_ID_RAZER_MAMBA_WIRELESS_WIRED:
+    case USB_DEVICE_ID_RAZER_VIPER_ULTIMATE_WIRELESS:
+    case USB_DEVICE_ID_RAZER_DEATHADDER_V2_PRO_WIRELESS:
+    case USB_DEVICE_ID_RAZER_VIPER_ULTIMATE_WIRED:
+    case USB_DEVICE_ID_RAZER_DEATHADDER_V2_PRO_WIRED:
+    case USB_DEVICE_ID_RAZER_COBRA_PRO:
+    case USB_DEVICE_ID_RAZER_ATHERIS_RECEIVER:
+    case USB_DEVICE_ID_RAZER_BASILISK_X_HYPERSPEED:
+    case USB_DEVICE_ID_RAZER_OROCHI_V2_RECEIVER:
+    case USB_DEVICE_ID_RAZER_OROCHI_V2_BLUETOOTH:
+    case USB_DEVICE_ID_RAZER_PRO_CLICK_RECEIVER:
+    case USB_DEVICE_ID_RAZER_PRO_CLICK_WIRED:
+    case USB_DEVICE_ID_RAZER_DEATHADDER_V2_X_HYPERSPEED:
+    case USB_DEVICE_ID_RAZER_VIPER_V2_PRO_WIRED:
+    case USB_DEVICE_ID_RAZER_VIPER_V2_PRO_WIRELESS:
+    case USB_DEVICE_ID_RAZER_DEATHADDER_V3_PRO_WIRED:
+    case USB_DEVICE_ID_RAZER_DEATHADDER_V3_PRO_WIRELESS:
+    case USB_DEVICE_ID_RAZER_PRO_CLICK_MINI_RECEIVER:
+    case USB_DEVICE_ID_RAZER_VIPER_V3_HYPERSPEED:
+    case USB_DEVICE_ID_RAZER_VIPER_MINI_SE_WIRELESS:
+    case USB_DEVICE_ID_RAZER_VIPER_MINI_SE_WIRED:
+    case USB_DEVICE_ID_RAZER_HYPERPOLLING_WIRELESS_DONGLE:
+    case USB_DEVICE_ID_RAZER_NAGA_V2_HYPERSPEED_RECEIVER:
+        return 1;
+    default:
+        return 0;
+    }
+}
+
+int razer_scale_value(int value, int minIn, int maxIn, int minOut, int maxOut)
+{
+    // Ensure the input value is within the valid range.
+    if (value < minIn) value = minIn;
+    if (value > maxIn) value = maxIn;
+
+    return ((value - minIn) * (maxOut - minOut)) / (maxIn - minIn) + minOut;
+}
+
+static int razer_battery_get_property(struct power_supply *ps,
+                                      enum power_supply_property property, union power_supply_propval *val)
+{
+    struct razer_mouse_device *dev = power_supply_get_drvdata(ps);
+
+    int ret = 0;
+    switch (property) {
+    case POWER_SUPPLY_PROP_PRESENT:
+        val->intval = 1;
+        break;
+    case POWER_SUPPLY_PROP_SCOPE:
+        val->intval = POWER_SUPPLY_SCOPE_DEVICE;
+        break;
+    case POWER_SUPPLY_PROP_CAPACITY:
+        val->intval = razer_scale_value(razer_read_charge_level(dev), 0, 255, 0, 100);
+        break;
+    case POWER_SUPPLY_PROP_MODEL_NAME:
+        val->strval = razer_get_device_type(dev);
+        break;
+    case POWER_SUPPLY_PROP_SERIAL_NUMBER:
+        // this is the random serial number because we don't want upower to think the wired and wireless devices are the same
+        val->strval = dev->serial;
+        break;
+    case POWER_SUPPLY_PROP_ONLINE:
+        val->intval = 1;
```

Comment:
I think 'online' should be whether it's plugged in or not so this should be also dependent on razer_read_charge_status?

Reply:
Or actually it seems most drivers upstream in the kernel don't even implement this prop?

Reply:
I can't find clear documentation on it so I'm not even sure what it means or what the difference between online and present is but it does appear to be properly implemented in xpadneo.
https://github.com/atar-axis/xpadneo/blob/master/hid-xpadneo/src/hid-xpadneo.c#L534

Reply:
Please check in upstream kernel what props are used for this use case, since the Razer devices are quite simple with not too much things it should be easy to find something equivalent there.

Reply:
According to official sysfs docs online means means vbus is present aka plugged in. I assume there is no direct way to detect that so I just changed it to return charge status.